### ISSUE TITLE
Add custom variable to annotations

### ISSUE THREAD

**Issue by LouisSnz** [Unity-Technologies/com.unity.perception#518]
Hi,
I am trying to figure out how to add a custom variable to the object annotations but I do not understand what I should change to do so. For now there is _label_id, label_name, instance_id, x, y, width, height_ and I would like to add the distance between the object and the camera.
Thank you in advance.

**Comment by sleal-unity** [comment]
Hi @LouisSnz,

Below I've provided a few c# classes that together create a custom labeler that can output the distance for all labeled objects in the scene. To use this labeler, please copy the following 4 scripts into your project's scripts folder. The name of each script file should be the name of each respective class it contains.

------------------------------------------

This ObjectDistanceLabeler does the work of creating a new distance annotation for each labeled object in the scene.
```csharp
using System;
using UnityEngine;
using UnityEngine.Perception.GroundTruth;
using UnityEngine.Perception.GroundTruth.DataModel;
using UnityEngine.Rendering;

[Serializable]
public class ObjectDistanceLabeler : CameraLabeler
{
    Camera m_Camera;
    AnnotationDefinition m_AnnotationDefinition;

    public string annotationId = "ObjectDistance";
        
    public override string description => "Returns the distance of each labeled object from the camera every frame.";

    public override string labelerId => annotationId;
        
    protected override bool supportsVisualization => false;

    protected override void Setup()
    {
        m_Camera = perceptionCamera.attachedCamera;
        m_AnnotationDefinition = new ObjectDistanceAnnotationDefinition(annotationId);
        DatasetCapture.RegisterAnnotationDefinition(m_AnnotationDefinition);
    }
        
    protected override void OnBeginRendering(ScriptableRenderContext ctx)
    {
        var annotation = new ObjectDistanceAnnotation(m_AnnotationDefinition, sensorHandle.Id);

        foreach (var labeling in LabelManager.singleton.registeredLabels)
        {
            annotation.distanceInfos.Add(new ObjectDistanceInfo
            {
                instanceId = (int)labeling.instanceId,
                distance = (labeling.transform.position - m_Camera.transform.position).magnitude
            });
        }
            
        perceptionCamera.SensorHandle.ReportAnnotation(m_AnnotationDefinition, annotation);
    }
}
```

This ObjectDistanceAnnotation class defines how a collection of distance values can be written to the dataset,
```csharp
using System;
using System.Collections.Generic;
using UnityEngine.Perception.GroundTruth.DataModel;

public class ObjectDistanceAnnotation : Annotation
{
    public List<ObjectDistanceInfo> distanceInfos = new List<ObjectDistanceInfo>();
        
    public ObjectDistanceAnnotation(AnnotationDefinition definition, string sensorId)
        : base(definition, sensorId) { }
    
    public override void ToMessage(IMessageBuilder builder)
    {
        base.ToMessage(builder);

        foreach (var info in distanceInfos)
        {
            var nested = builder.AddNestedMessageToVector("instances");
            nested.AddInt("instanceId", info.instanceId);
            nested.AddFloat("distance", info.distance);
        }
    }
}
```

This ObjectDistanceAnnotationDefinition class provides some meta information about the distance annotations written in the dataset.
```csharp
using System;
using UnityEngine.Perception.GroundTruth.DataModel;

public class ObjectDistanceAnnotationDefinition : AnnotationDefinition
{
    public override string modelType => "type.unity.com/unity.solo.ObjectDistanceAnnotation";

    public override string description => "Returns the distance of each labeled object from the camera every frame.";
        
    public ObjectDistanceAnnotationDefinition(string id) : base(id) { }
}
```

And finally this ObjectDistanceInfo struct encapsulates the each labeled object's captured distance information.
```csharp
using System;

public struct ObjectDistanceInfo
{
    public int instanceId;
    public float distance;
}
```

**Comment by LouisSnz** [comment]
Thank you very much for your answer @sleal-unity, it might save me. 
I am pretty sure the solution you proposed would work for me if I add it on top of the BoundingBox2D labeler and I merge the data after. However, I would like ideally to have a single output file for both the 2D Bounding Box and the distance of the object (example, as follow) :
```json
"id": "BoundingBox",
"annotation_definition": "BoundingBox",
"values": [
   {
   "label_id": 1,
   "label_name": "Car",
    "instance_id": 2,
    "x": 302.0,
    "y": 280.0,
    "width": 30.0,
    "height": 10.0,
    "distance": 50.0
    },
```
So far, I tried to merge what you gave with with the existing BoundingBox Labeler but my last problem is to get my object position in the BoundingBoxLabeler.cs file.

```csharp
void OnRenderedObjectInfosCalculated(int frameCount, NativeArray<RenderedObjectInfo> renderedObjectInfos)
{
    if (!m_AsyncData.TryGetValue(frameCount, out var asyncData))
        return;

    m_AsyncData.Remove(frameCount);
    using (s_BoundingBoxCallback.Auto())
    {
        var bbValues = new List<BoundingBox>();
        for (var i = 0; i < renderedObjectInfos.Length; i++)
        {
            var objectInfo = renderedObjectInfos[i];
            if (!asyncData.labelEntryMatchCache.TryGetLabelEntryFromInstanceId(objectInfo.instanceId, out var labelEntry, out _))
                continue;

            bbValues.Add(new BoundingBox
                {
                    labelId = labelEntry.id,
                    labelName = labelEntry.label,
                    instanceId = (int)objectInfo.instanceId,
                    origin = new Vector2(objectInfo.boundingBox.x, objectInfo.boundingBox.y),
                    dimension = new Vector2(objectInfo.boundingBox.width, objectInfo.boundingBox.height),

                    // Here is what I've tried, but it doesn't work
                    distance = (objectInfo.transform.position - m_Camera.transform.position).magnitude  
                }
            );
        }

        if (!PerceptionCamera.useAsyncReadbackIfSupported && frameCount != Time.frameCount)
            Debug.LogWarning("Not on current frame: " + frameCount + "(" + Time.frameCount + ")");

        m_ToVisualize = bbValues;

        boundingBoxesCalculated?.Invoke(new BoundingBoxesCalculatedEventArgs()
        {
            data = bbValues,
            frameCount = frameCount
        });

        var toReport = new BoundingBoxAnnotation(m_AnnotationDefinition, perceptionCamera.ID, bbValues);
        asyncData.annotation.Report(toReport);
        asyncData.labelEntryMatchCache.Dispose();
    }
}
```
Would you have any idea on how to correct that?

**Comment by sleal-unity** [comment]
Hi again @LouisSnz,

TLDR: Unfortunately the bounding box 2D labeler subscribes to a special "RenderedObjectInfosCalculated" event that doesn't quite provide the 3D positional data you would need to perform the distance calculation you're interested in.

To make a short story long (lol), this "RenderedObjectInfosCalculated" event does some work in the background to readback the instance segmentation image generated on the GPU so that the CPU can look at the pixels of the various objects in the segmentation image to calculate their 2D bounding boxes. After these calculations, the "RenderedObjectInfosCalculated" event is invoked to return the calculated 2D bounding box data or "rendered object infos". So that "objectInfo" variable you've noted in the code snippet you sent is really just a tiny struct holding some 2D bounding box data + some metadata for one visible object in the current frame. Unfortunately, you cannot access that object's 3D transform information through this struct.

In order to obtain the 3D object transform information you need to perform your desired distance calculation, you will have to work around the existing "RenderedObjectInfosCalculated" event. Therefore, my original suggestion was simply to create a separate labeler to output this distance information more directly. It does have the disadvantage to appearing in a less convenient spot in the dataset, but hopefully your dataset consumer code is flexible enough to adjust for this.

I hope this wall of text provides some context as to why it is challenging to modify the 2D bounding box labeler to get the distance calculation to appear inside of the existing annotation.

**Comment by LouisSnz** [comment]
Thanks for this clear answer @sleal-unity.
I actually tried to add the _ObjectDistanceLabeler_ to my camera as you suggested me, but it seems that it is not synchronised with the _BoundingBoxLabeler_. When I look at my results file, on some frames I have only distance annotations or only 2D BB.
I'll try to explore on creating my own annotation using the _ RenderedObjectInfosCalculated_ event, and merge both distance and BB. 
I also though about a workaround using the center's transform of the 3D BB labeler, and then calculate the distance with the translation of the 2D BB labeler. It's not ideal but it could do the job.

**Comment by StevenBorkman** [comment]
I'm going to close this issue for now, but feel free to re-open it or create a new one if you need more assistance.
